File: 3661-maximum-walls-destroyed-by-robots/3661-maximum-walls-destroyed-by-robots.py
class Solution:
    def maxWalls(self, robots: List[int], distance: List[int], walls: List[int]) -> int:
        A = sorted(zip(robots, distance))
        N = len(A)
        walls.sort()

        def count(L, R):
            if L > R:
                return 0
            return bisect_right(walls, R) - bisect_left(walls, L)

        avail = 0
        used = count(A[0][0] - A[0][1], A[0][0] - 1)
        A.append([inf, 0])
        for i in range(N):
            l, dl = A[i]
            r, dr = A[i + 1]

            l1, r1 = l + 1, min(l + dl, r - 1)
            l2, r2 = max(r - dr, l + 1), r - 1
            
            left = count(l1, r1)
            right = count(l2, r2)
            both = left + right - count(max(l1, l2), min(r1, r2))

            navail = max(avail + left, used)
            nused = max(avail + both, used + right)
            avail, used = navail, nused

        for x in set(x for x,_ in A):
            used += count(x, x)
        return used

File: 3661-maximum-walls-destroyed-by-robots/3661-maximum-walls-destroyed-by-robots.cpp
class Solution {
public:
    int maxWalls(vector<int>& robots, vector<int>& d, vector<int>& walls) {
        int n = robots.size();
        vector<array<int,2>> x(n);
        for(int i=0; i<n; i++) x[i][0] = robots[i],x[i][1] = d[i];
        sort(walls.begin(),walls.end());
        sort(x.begin(),x.end());
        x.push_back({(int)1e9,0});

        // finds the no of walls in the range [l,r];
        auto query=[&](int l,int r){
            if(l>r) return 0;
            auto it1 = upper_bound(walls.begin(),walls.end(),r);
            auto it2 = lower_bound(walls.begin(),walls.end(),l);
            int res = it1-it2;
            return res;
        };
        
        vector<array<int,2>> dp(n);
        // base case
        dp[0][0] = query(x[0][0] - x[0][1],x[0][0]);
        if(n>1) dp[0][1] = query(x[0][0],min(x[1][0]-1, x[0][0]+x[0][1]));
        else dp[0][1] = query(x[0][0], x[0][0] + x[0][1]);

        //transition
        for(int i=1; i<n; i++){
            dp[i][1] = max(dp[i-1][1],dp[i-1][0]) + query(x[i][0],min(x[i+1][0]-1,x[i][0]+x[i][1]));
            
            dp[i][0] = dp[i-1][0]+ query(max(x[i][0]-x[i][1],x[i-1][0]+1),x[i][0]);
            int res = dp[i-1][1] 
                        + query(max(x[i][0]-x[i][1],x[i-1][0]+1),x[i][0]) 
                        - query(max(x[i][0]-x[i][1],x[i-1][0]+1),min(x[i-1][0]+x[i-1][1],x[i][0]-1));
            dp[i][0] = max(res,dp[i][0]);
        }
        return max(dp[n-1][0],dp[n-1][1]);
    }
};